File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.py
def solution(new_id):
    answer = ''
    
    # 1단계
    new_id = new_id.lower();
    print('1:', new_id)

    # 2단계
    new_id = list(new_id)
    for i in range(len(new_id)):
        tmp = new_id[i]
        if tmp.isalnum() or tmp == '-' or tmp == '_' or tmp == '.':
            continue
        else:
            new_id[i] = ''
    new_id = ''.join(new_id)
    print('2:', new_id)

    # 3단계
    stack = []
    for i in range(len(new_id)):
        # stack에 값이 있고, 현재 값이 '.'이고 직전 값이랑 같으면 현재 값은 저장하지 않고 넘어간다
        # 즉, 첫 '.'만 남기고 그 이후 연속된 '.'들은 모두 무시
        tmp = new_id[i]
        if stack and tmp == '.' and stack[-1] == tmp:
            continue
        else:
            stack.append(tmp)
    new_id = ''.join(stack)
    print('3:', new_id)
    
    # 4단계
    if new_id != '':
        if new_id[0] == '.' and new_id[-1] == '.':
            new_id = new_id[1:-1]
        elif new_id[0] == '.':
            new_id = new_id[1:]
        elif new_id[-1] == '.':
            new_id = new_id[:-1]
        
    print('4:', new_id)
    
    # 5단계
    if new_id == '':
        new_id = 'a'
    print('5:', new_id)
    
    # 6단계
    if len(new_id) > 15:
        if new_id[14] == '.':
            new_id = new_id[:14]
        else:
            new_id = new_id[:15]
    print('6:', new_id)
    
    # 7단계
    if len(new_id) <= 2:
        new_id = list(new_id)
        last_ch = new_id[-1]
        while len(new_id) < 3:
            new_id.append(last_ch)
        new_id = ''.join(new_id)
    print('7:', new_id)
    
    return new_id

File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.java
class Solution {
    public String solution(String new_id) {
        // 길이 : 3~15
        // 소문자, 숫자, -, _, . 만 사용 가능
        // .는 맨 앞, 맨 뒤, 연속 사용 불가능
        
        /*
        1. 소문자로
        2. 소문자, 숫자, -, _, . 제외한 문자 제거
        3. 연속된 .. 들 모두 .으로 
        4. 맨 앞, 맨 뒤의 . 제거
        5. 빈 문자열 -> a 대입
        6. 길이 16 이상 -> 15 초과 문자 모두 제거 (제거 후, 마지막이 .이라면 이것도 제거)
        7. 길이 2 이하 -> 마지막 문자를 길이 3 될 때까지 끝에 붙임
        */
        
        String answer = "";
        
        // 1단계
        answer = new_id.toLowerCase();
        
        // 2단계
        answer = replaces(answer);
        
        // 3단계
        int len = answer.length();
        while(true) {
            answer = answer.replace("..", ".");
            
            if (answer.length() == len) break;
            len = answer.length();
        }
        
        // 4단계
        if (!answer.isEmpty() && answer.charAt(0) == '.') answer = answer.substring(1, len);
        len = answer.length();
        if (!answer.isEmpty() && answer.charAt(len-1) == '.') answer = answer.substring(0, len-1);
        
        // 5단계
        if (answer.isEmpty()) answer = "a";
        
        // 6단계
        len = answer.length();
        if (len >= 16) {
            answer = answer.substring(0, 15);
            if (answer.charAt(14) == '.') answer = answer.substring(0, 14);
        }
        
        // 7단계
        if (len <= 2) {
            StringBuilder sb = new StringBuilder();
            sb.append(answer);
            char end = answer.charAt(len-1);
            while (sb.length() < 3) {
                sb.append(end);
            }
            answer = sb.toString();
        }
        return answer;
    }
    
    static String replaces(String newId) {
        int n = newId.length();
        for (int i = 0; i < n; i++) {
            char cur = newId.charAt(i);
            // 소문자, 숫자, -, _, . 인 경우 pass
            if ((cur >= 97 && cur <= 122) || (cur >= 48 && cur <= 57) || cur == '-' || cur == '_' || cur == '.') continue;
            newId = newId.replace(cur, ' ');
        }
        
        newId = newId.replaceAll(" ", "");
        return newId;
    }
    
}